File: examples/python/parallel_tools.py
"""Deterministic read-only batch example, with no Provider or external service."""
import asyncio
import json
from purra.contracts import ExecutionState, ToolBatchRequest, ToolCall, ToolExecutionLimits, ToolHandlerResult, ToolPolicy, ToolSchema
from purra.ports import ToolRegistration
from purra.tools import CoreToolExecutor, InMemoryToolCatalog


async def main():
    records = {'alpha': {'count': 2}, 'beta': {'count': 3}}
    async def read(state, arguments, signal=None):
        return ToolHandlerResult(json.dumps(records[arguments['key']]), effect_state='not_started')
    async def scope(state, arguments, signal=None):
        return None if arguments['key'] in records else 'Unknown record'
    class Sink:
        async def emit(self, event):
            print(event.type, event.payload['toolCallId'])
    registration = ToolRegistration(
        ToolSchema('readRecord', 'Read an approved local record', {
            'type': 'object', 'properties': {'key': {'type': 'string'}},
            'required': ['key'], 'additionalProperties': False,
        }), read, ToolPolicy(mode='read', title='Read record'),
        scope_validator=scope, concurrency_safe=True,
    )
    executor = CoreToolExecutor(InMemoryToolCatalog((registration,)),
        limits=ToolExecutionLimits(max_concurrency=2))
    batch = ToolBatchRequest(None, tuple(
        ToolCall(key, 'readRecord', json.dumps({'key': key})) for key in records
    ), frozenset({'readRecord'}), ExecutionState())
    result = await executor.execute_batch(batch, Sink())
    assert [row.tool_call_id for row in result.results] == list(records)
    print([json.loads(row.content) for row in result.results])


if __name__ == '__main__':
    asyncio.run(main())

File: typescript/examples/parallel-tools.ts
// Deterministic gateway and local records; no Provider or external service.
import { Agent, type ModelCapabilitySnapshot, type ModelGateway } from "purra";

const capabilities: ModelCapabilitySnapshot = {
  schemaVersion: 2,
  profileId: "example:parallel-tools",
  providerProtocol: "custom",
  contextWindowTokens: 16_000,
  maxGenerationTokens: 512,
  thinkingTokenAccounting: "unknown",
  protocol: {
    reasoningControl: "selectable", reasoningReplay: "ignored", toolCalling: "supported",
    requiredToolChoice: "supported", parallelToolCalls: "supported", streaming: "unavailable",
    cancellation: "supported", assistantContentWithToolCalls: "optional", jsonSchemaLevel: "unknown",
    streamFinishSemantics: "normalized", usageSemantics: "normalized",
  },
};

const records: Readonly<Record<string, {count: number}>> = {alpha:{count:2},beta:{count:3}};
let round = 0;
const model: ModelGateway = {
  capabilities,
  async invoke(request) {
    round++;
    return {appliedGenerationLimit:request.outputBudget.maxGenerationTokens,
      message:round===1 ? {role:"assistant",content:"",toolCalls:Object.keys(records).map(key=>({id:key,name:"readRecord",arguments:{key}}))}
        : {role:"assistant",content:"Read both approved records."},
      finishReason:round===1 ? "tool_calls" : "stop"};
  },
};
const agent = new Agent({model,toolLimits:{maxConcurrency:2},tools:[{
  name:"readRecord",description:"Read an approved local record",
  inputSchema:{type:"object",properties:{key:{type:"string"}},required:["key"],additionalProperties:false},
  policy:{mode:"read",title:"Read record"},concurrencySafe:true,
  scope:args=>Object.hasOwn(records,(args as {key:string}).key) ? undefined : "Unknown record",
  async run(args) {return {content:records[(args as {key:string}).key]!,effectState:"not_started"};},
}]});
const result = await agent.invoke({messages:[{role:"user",content:"Read alpha and beta"}]});
console.log(result.messages.filter(message=>message.role==="tool").map(message=>message.toolCallId));
console.log(result.output);
